xfrd.c

```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
 		xfrd_process_zonestat_inc_task(xfrd, task);
 		break;
 #endif
+	case task_add_catzone: {
+		char* zname = (char*)task->zname;
+		char* pattern = zname + strlen(zname) + 1;
+		char* catname = pattern + strlen(pattern) + 1;
+		char* member_id = catname + strlen(catname) + 1;
+
+		struct zone_options* zopt = NULL;
+		pattern_options_type* patopt;
+
+		log_msg(LOG_INFO, "now adding catzone %s %s %s %s", zname, pattern, catname, member_id);
+
+		patopt = pattern_options_find(xfrd->nsd->options, pattern);
+		if (!patopt) {
+			patopt = pattern_options_create(xfrd->region);
+			patopt->pname = pattern;
+			pattern_options_add_modify(xfrd->nsd->options, patopt);
+		} 
+		zopt = zone_options_find(xfrd->nsd->options, dname_parse(xfrd->region, zname));
+		if (!zopt) {
+			zopt = zone_list_zone_insert(xfrd->nsd->options, zname, pattern, 0, 0);
+		} else {
+			log_msg(LOG_INFO, "catzone already added %s", zname);
+			return;
+		}
+		task_new_add_catzone(xfrd->nsd->task[xfrd->nsd->mytask],
```

Comment:
When xfrd receives a catzone addition task as a taskresult from reload, it then makes a new task for the next reload to add the catzone? Is that intentionally the work order, or a loop?

Reply:
Because this is called from xfrd, the result of the new task ends up at difffile rather than back here, so there is no loop (or there should not be). Admittedly this part still confuses me as well.